File: parallel_merge_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <thread>
#include <future>
#include <random>
#include <chrono>
#include <cmath>
#include <memory>

class ParallelMergeSort {
private:
    static const int SEQUENTIAL_THRESHOLD = 1000; // 串行阈值
    static const int MAX_DEPTH = 4;             // 最大递归深度

    // 串行归并排序
    static void sequential_merge_sort(std::vector<int>& arr, int left, int right) {
        if (left >= right) return;

        if (right - left + 1 <= SEQUENTIAL_THRESHOLD) {
            std::sort(arr.begin() + left, arr.begin() + right + 1);
            return;
        }

        int mid = left + (right - left) / 2;
        sequential_merge_sort(arr, left, mid);
        sequential_merge_sort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }

    // 归并操作
    static void merge(std::vector<int>& arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        std::vector<int> left_arr(n1), right_arr(n2);

        // 复制数据到临时数组
        for (int i = 0; i < n1; i++)
            left_arr[i] = arr[left + i];
        for (int j = 0; j < n2; j++)
            right_arr[j] = arr[mid + 1 + j];

        // 归并临时数组
        int i = 0, j = 0, k = left;
        while (i < n1 && j < n2) {
            if (left_arr[i] <= right_arr[j]) {
                arr[k] = left_arr[i];
                i++;
            } else {
                arr[k] = right_arr[j];
                j++;
            }
            k++;
        }

        // 复制剩余元素
        while (i < n1) {
            arr[k] = left_arr[i];
            i++;
            k++;
        }
        while (j < n2) {
            arr[k] = right_arr[j];
            j++;
            k++;
        }
    }

    // 递归并行归并排序
    static void parallel_merge_sort_recursive(std::vector<int>& arr, int left, int right, int depth) {
        if (left >= right || depth >= MAX_DEPTH) {
            sequential_merge_sort(arr, left, right);
            return;
        }

        int mid = left + (right - left) / 2;

        // 并行处理左右两部分
        auto future1 = std::async(std::launch::async, [&]() {
            parallel_merge_sort_recursive(arr, left, mid, depth + 1);
        });

        auto future2 = std::async(std::launch::async, [&]() {
            parallel_merge_sort_recursive(arr, mid + 1, right, depth + 1);
        });

        // 等待两个子任务完成
        future1.wait();
        future2.wait();

        // 合并结果
        merge(arr, left, mid, right);
    }

public:
    // 并行归并排序主函数
    static void sort(std::vector<int>& arr) {
        if (arr.size() <= SEQUENTIAL_THRESHOLD) {
            std::sort(arr.begin(), arr.end());
            return;
        }
        parallel_merge_sort_recursive(arr, 0, arr.size() - 1, 0);
    }

    // 串行归并排序（用于性能对比）
    static void sequential_sort(std::vector<int>& arr) {
        sequential_merge_sort(arr, 0, arr.size() - 1);
    }
};

// 性能测试工具
class PerformanceTimer {
private:
    std::chrono::high_resolution_clock::time_point start_time;

public:
    void start() {
        start_time = std::chrono::high_resolution_clock::now();
    }

    double elapsed_seconds() {
        auto end_time = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end_time - start_time);
        return duration.count() / 1000000.0;
    }
};

// 生成随机数据
std::vector<int> generate_random_data(size_t size) {
    std::vector<int> data(size);
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(1, 1000000);

    for (size_t i = 0; i < size; i++) {
        data[i] = dis(gen);
    }
    return data;
}

// 验证排序正确性
bool is_sorted(const std::vector<int>& arr) {
    for (size_t i = 1; i < arr.size(); i++) {
        if (arr[i] < arr[i-1]) return false;
    }
    return true;
}

// 性能基准测试
void benchmark_sorting_algorithms(size_t array_size = 100000, int num_tests = 3) {
    std::cout << "=== 并行归并排序性能测试 ===" << std::endl;
    std::cout << "数组大小: " << array_size << std::endl;
    std::cout << "CPU核心数: " << std::thread::hardware_concurrency() << std::endl;
    std::cout << std::endl;

    // 生成测试数据
    std::vector<int> test_data = generate_random_data(array_size);

    // 测试策略
    struct TestStrategy {
        std::string name;
        void (*sort_func)(std::vector<int>&);
    };

    std::vector<TestStrategy> strategies = {
        {"串行归并排序", ParallelMergeSort::sequential_sort},
        {"并行归并排序", ParallelMergeSort::sort}
    };

    std::vector<double> results;

    for (const auto& strategy : strategies) {
        std::cout << "测试: " << strategy.name << std::endl;

        std::vector<double> times;
        for (int i = 0; i < num_tests; i++) {
            std::vector<int> data_copy = test_data;

            PerformanceTimer timer;
            timer.start();

            strategy.sort_func(data_copy);

            double execution_time = timer.elapsed_seconds();
            times.push_back(execution_time);

            // 验证排序正确性
            if (!is_sorted(data_copy)) {
                std::cout << "  ❌ 排序结果不正确!" << std::endl;
                return;
            }
        }

        double avg_time = 0.0;
        for (double time : times) avg_time += time;
        avg_time /= times.size();

        double min_time = *std::min_element(times.begin(), times.end());
        results.push_back(avg_time);

        std::cout << "  平均时间: " << avg_time << "s" << std::endl;
        std::cout << "  最佳时间: " << min_time << "s" << std::endl;
        std::cout << std::endl;
    }

    // 计算加速比
    std::cout << "=== 加速比分析 ===" << std::endl;
    double sequential_time = results[0];

    for (size_t i = 1; i < results.size(); i++) {
        double speedup = sequential_time / results[i];
        double efficiency = speedup / std::thread::hardware_concurrency();

        std::cout << strategies[i].name << ":" << std::endl;
        std::cout << "  加速比: " << speedup << "x" << std::endl;
        std::cout << "  效率: " << (efficiency * 100) << "%" << std::endl;
        std::cout << std::endl;
    }
}

// 演示函数
void demo_parallel_merge_sort() {
    std::cout << "=== 并行归并排序演示 ===" << std::endl;

    // 小规模数据演示
    std::vector<int> small_data = {64, 34, 25, 12, 22, 11, 90, 5, 77, 30};
    std::cout << "原始数据: ";
    for (int val : small_data) {
        std::cout << val << " ";
    }
    std::cout << std::endl;

    ParallelMergeSort::sort(small_data);

    std::cout << "排序结果: ";
    for (int val : small_data) {
        std::cout << val << " ";
    }
    std::cout << std::endl;

    bool is_correct = is_sorted(small_data);
    std::cout << "验证正确: " << (is_correct ? "✅" : "❌") << std::endl;
    std::cout << std::endl;

    // 性能测试
    benchmark_sorting_algorithms();
}

int main() {
    demo_parallel_merge_sort();
    return 0;
}

File: parallel_merge_sort.py
#!/usr/bin/env python3
"""
并行归并排序算法实现
基于第三章的分治法并行算法设计
"""

import time
import random
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor
import multiprocessing as mp

def sequential_merge_sort(arr):
    """串行归并排序实现"""
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = sequential_merge_sort(arr[:mid])
    right = sequential_merge_sort(arr[mid:])

    return merge(left, right)

def merge(left, right):
    """归并两个已排序的数组"""
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    return result

def parallel_merge_sort_threadpool(arr, max_workers=None, threshold=1000):
    """使用线程池的并行归并排序"""
    if len(arr) <= threshold:
        return sequential_merge_sort(arr)

    if max_workers is None:
        max_workers = mp.cpu_count()

    def _parallel_merge_sort(arr, depth=0, max_depth=4):
        if len(arr) <= threshold or depth >= max_depth:
            return sequential_merge_sort(arr)

        mid = len(arr) // 2
        left_part = arr[:mid]
        right_part = arr[mid:]

        with ThreadPoolExecutor(max_workers=2) as executor:
            left_future = executor.submit(_parallel_merge_sort, left_part, depth + 1, max_depth)
            right_future = executor.submit(_parallel_merge_sort, right_part, depth + 1, max_depth)

            left_sorted = left_future.result()
            right_sorted = right_future.result()

        return merge(left_sorted, right_sorted)

    return _parallel_merge_sort(arr)

def parallel_merge_sort_processpool(arr, threshold=1000):
    """使用进程池的并行归并排序"""
    if len(arr) <= threshold:
        return sequential_merge_sort(arr)

    def _merge_sort_worker(arr_chunk):
        """进程池工作函数"""
        return sequential_merge_sort(arr_chunk)

    def _parallel_merge_sort(arr, depth=0, max_depth=3):
        if len(arr) <= threshold or depth >= max_depth:
            return sequential_merge_sort(arr)

        mid = len(arr) // 2
        left_part = arr[:mid]
        right_part = arr[mid:]

        with ProcessPoolExecutor(max_workers=2) as executor:
            left_future = executor.submit(_parallel_merge_sort, left_part, depth + 1, max_depth)
            right_future = executor.submit(_parallel_merge_sort, right_part, depth + 1, max_depth)

            left_sorted = left_future.result()
            right_sorted = right_future.result()

        return merge(left_sorted, right_sorted)

    return _parallel_merge_sort(arr)

class ParallelMergeSort:
    """并行归并排序类，支持多种并行策略"""

    def __init__(self, strategy='threadpool', threshold=1000):
        self.strategy = strategy
        self.threshold = threshold
        self.cpu_count = mp.cpu_count()

    def sort(self, arr):
        """根据策略选择排序方法"""
        if self.strategy == 'sequential':
            return sequential_merge_sort(arr)
        elif self.strategy == 'threadpool':
            return parallel_merge_sort_threadpool(arr, threshold=self.threshold)
        elif self.strategy == 'processpool':
            return parallel_merge_sort_processpool(arr, threshold=self.threshold)
        else:
            raise ValueError(f"Unsupported strategy: {self.strategy}")

def benchmark_sorting_algorithms(arr_size=100000, num_tests=3):
    """性能基准测试"""
    print(f"=== 并行归并排序性能测试 ===")
    print(f"数组大小: {arr_size:,}")
    print(f"CPU核心数: {mp.cpu_count()}")
    print()

    # 生成测试数据
    test_data = [random.randint(1, 1000000) for _ in range(arr_size)]

    strategies = [
        ('串行归并排序', 'sequential', None),
        ('线程池并行排序', 'threadpool', None),
        ('进程池并行排序', 'processpool', None),
    ]

    results = {}

    for name, strategy, workers in strategies:
        print(f"测试: {name}")
        times = []

        for i in range(num_tests):
            data_copy = test_data.copy()

            start_time = time.time()
            sorter = ParallelMergeSort(strategy=strategy)
            sorted_data = sorter.sort(data_copy)
            end_time = time.time()

            execution_time = end_time - start_time
            times.append(execution_time)

            # 验证排序正确性
            is_sorted = all(sorted_data[i] <= sorted_data[i+1] for i in range(len(sorted_data)-1))
            if not is_sorted:
                print(f"  ❌ 排序结果不正确!")
                return

        avg_time = sum(times) / len(times)
        min_time = min(times)
        results[strategy] = avg_time

        print(f"  平均时间: {avg_time:.4f}s")
        print(f"  最佳时间: {min_time:.4f}s")
        print()

    # 计算加速比
    print("=== 加速比分析 ===")
    sequential_time = results['sequential']

    for name, strategy, _ in strategies:
        if strategy != 'sequential':
            speedup = sequential_time / results[strategy]
            efficiency = speedup / mp.cpu_count()
            print(f"{name}:")
            print(f"  加速比: {speedup:.2f}x")
            print(f"  效率: {efficiency:.2%}")
            print()

def demo_parallel_merge_sort():
    """演示并行归并排序"""
    print("=== 并行归并排序演示 ===")

    # 小规模数据演示
    small_data = [64, 34, 25, 12, 22, 11, 90, 5, 77, 30]
    print(f"原始数据: {small_data}")

    sorter = ParallelMergeSort(strategy='threadpool')
    sorted_data = sorter.sort(small_data.copy())

    print(f"排序结果: {sorted_data}")
    print(f"验证正确: {'✅' if sorted(small_data) == sorted_data else '❌'}")
    print()

    # 性能测试
    benchmark_sorting_algorithms()

if __name__ == "__main__":
    demo_parallel_merge_sort()